grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancer2Test.java

```diff
@@ -0,0 +1,844 @@
+import java.net.SocketAddress;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+
+/** Unit tests for {@link GrpclbLoadBalancer2}. */
+@RunWith(JUnit4.class)
+public class GrpclbLoadBalancer2Test {
+  private static final Attributes.Key<String> RESOLUTION_ATTR =
+      Attributes.Key.of("resolution-attr");
+  private static final String SERVICE_AUTHORITY = "api.google.com";
+
+  private static final MethodDescriptor<String, String> TRASH_METHOD = MethodDescriptor.create(
+      MethodDescriptor.MethodType.UNARY, "/service/trashmethod",
+      new StringMarshaller(), new StringMarshaller());
+
+  private static class StringMarshaller implements Marshaller<String> {
+    static final StringMarshaller INSTANCE = new StringMarshaller();
+
+    @Override
+    public InputStream stream(String value) {
+      return new ByteArrayInputStream(value.getBytes(UTF_8));
+    }
+
+    @Override
+    public String parse(InputStream stream) {
+      try {
+        return new String(ByteStreams.toByteArray(stream), UTF_8);
+      } catch (IOException ex) {
+        throw new RuntimeException(ex);
+      }
+    }
+  }
+
+  @Mock
+  private Helper helper;
+  @Mock
+  private Subchannel mockSubchannel;
+  @Mock
+  private LoadBalancerGrpc.LoadBalancerImplBase mockLbService;
+  @Captor
+  private ArgumentCaptor<StreamObserver<LoadBalanceResponse>> lbResponseObserverCaptor;
+  private final LinkedList<StreamObserver<LoadBalanceRequest>> lbRequestObservers =
+      new LinkedList<StreamObserver<LoadBalanceRequest>>();
+  private final LinkedList<Subchannel> mockSubchannels = new LinkedList<Subchannel>();
+  private final LinkedList<ManagedChannel> fakeOobChannels = new LinkedList<ManagedChannel>();
+  private final ArrayList<Subchannel> subchannelTracker = new ArrayList<Subchannel>();
+  private final ArrayList<ManagedChannel> oobChannelTracker = new ArrayList<ManagedChannel>();
+  private final ArrayList<String> failingLbAuthorities = new ArrayList<String>();
+  private io.grpc.Server fakeLbServer;
+  @Captor
+  private ArgumentCaptor<SubchannelPicker> pickerCaptor;
+  private final ChannelExecutor channelExecutor = new ChannelExecutor();
+  private final Metadata headers = new Metadata();
+  @Mock
+  private LoadBalancer2.Factory pickFirstBalancerFactory;
+  @Mock
+  private LoadBalancer2 pickFirstBalancer;
+  @Mock
+  private LoadBalancer2.Factory roundRobinBalancerFactory;
+  @Mock
+  private LoadBalancer2 roundRobinBalancer;
+  private GrpclbLoadBalancer2 balancer;
+
+  @SuppressWarnings("unchecked")
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    when(pickFirstBalancerFactory.newLoadBalancer(any(Helper.class)))
+        .thenReturn(pickFirstBalancer);
+    when(roundRobinBalancerFactory.newLoadBalancer(any(Helper.class)))
+        .thenReturn(roundRobinBalancer);
+    doAnswer(new Answer<StreamObserver<LoadBalanceRequest>>() {
```

Comment:
why not implement it and then spy instead of mock and then `doAnswer`?
````
private final LoadBalancerGrpc.LoadBalancerImplBase mockLbService = 
    syp(new LoadBalancerGrpc.LoadBalancerImplBase {
      @Override
       public StreamObserver<LoadBalanceRequest> balanceLoad(StreamObserver<LoadBalanceResponse> rspOb) {
          ....
       }
    });
````